### ISSUE TITLE
ModMaker, DLC Mods, Extraction/TOCing

### ISSUE THREAD

**Issue by ghost** [ME3Explorer/ME3Explorer#132]
Just ran a .mod for the first time in probably 2 years. A few separate aspects related to the issue title.
1. Upon loading a .mod, ModMaker does not recognize DLC mods. Only BW DLC are detected. It would be a very good thing to roll DLC mod detection into this.
2. ModMaker Debug output indicates DLC extraction is undertaken upon running a job. In my case, all DLC were already extracted, so it wasn't a big deal. However, there is no warning for the user about this. One should likely be implemented, similar to Texplorer when treebuilding.
3. Even if a .mod file only edits base game PCCs, ALL DLCs are TOCed. This includes BW DLC and DLC mods. The latter is particularly funny, since ModMaker won't install jobs to these files, since it doesn't recognize them (see 1 above).

**Comment by KFreon** [comment]
As per skype:
1. Intentional. Sorry :(
2. and 3. This was due to the DLC checks not actually checking if DLC updates were required.
Message added to notify users that DLC extraction is about to take place.

**Comment by ghost** [comment]
@KFreon -- We still have issues with this that need to be worked out. Also see issue #140; same thing.

ModMaker is still TOCing all DLC -- including DLC mods -- even when the DLC isn't edited by the mod:

![mm_moretocing](https://cloud.githubusercontent.com/assets/13645804/13020883/aefe85a0-d18c-11e5-82e9-ee4a7eb8c5d8.PNG)

This job only affects the CE DLC. I didn't get a "DLC extraction" warning, which is good, since it was already extracted.

Then, I re-ran the job, after restoring my original SFAR. 

This time I also got no warning about the extraction... which isn't good.

Users need to be warned if we're going to extract their DLC, just like Texplorer does. In addition, if we're going to make extraction and TOCing of all DLC mandatory -- even when a DLC is not otherwise affected by the .mod file -- then the user absolutely **must** be warned before the tool alters their files. We have to let them know what they're getting into and give them the option to opt out.

I'm certainly open to suggestions and am open to whatever way creates the least work for you. Right now, I see two logical ways to do this:
1. We have a "first time setup" a little similar to Texplorer. When ModMaker is first opened it will: 1-check if DLC are already extracted, and 2-if not, prompt a message that warns the user that all will be extracted before they can use the tool. They should be able to abort without consequence if they wish. If the tool sees all DLC are already extracted, then there's no warning message. My guess is simply checking for a PCConsoleTOC.bin inside each DLC's root folder would be a foolproof way to ensure it's extracted.

Or....
1. Upon selecting _Run Selected/All_ -- but before modifying any files -- the tool simply checks if the .mod file mods any DLC. If it does, then it triggers a warning something like this:

"WARNING! You're about to apply a .mod job that edits DLC files. All DLC still packed in their original SFARs are about to be unpacked. The more DLC you have the longer it will take and the more hard drive space will be required. Click Okay to accept, or Cancel to abort."

The second option seems like the least work, and also the least prone to error. It also doesn't require checking the state of the DLC, specifically. It also allows you to keep the current extraction/TOC all scheme.

Like I said, I'm open to other suggestions, but we need to let players know that we're going to extract all their DLC and give them a chance to say, "No, this isn't okay." And, we need to do it before any of their files are modified.

**Comment by KFreon** [comment]
I remember why Fob did this. There was a weird issue where updating a single DLC toc doesn't work properly. 
Imma undo that and see if we can lock down that issue.

There's no real way to display a message across all tools properly, but if the single DLC update is fixed, this won't be a concern.

Now, here's the kicker. If the mod is looking for a DLC file and can't find it because it's in a packed DLC, it'll ignore it. There'll be a message in the DebugWindow, but otherwise, nada.

**Comment by ghost** [comment]
@KFreon -- Uh... okay slow down a minute before messing with DLC TOCing. 

> I remember why Fob did this. There was a weird issue where updating a single DLC toc doesn't work properly. Imma undo that and see if we can lock down that issue.

**Please** don't do that without consulting Fob and Heff. Remember that when I suggest things, I don't understand the full ramifications of what it means to the underlying code, what's linked together, what isn't, etc. To me, I'm suggesting what should happen inside ModMaker, _only_. 

Heff and Fob put those changes in place for a **very good reason** -- it's what finally fixed everything from fubaring with textures, BW DLC, and DLC Mods. Before they put those changes in place things were a complete mess, and screwing with it could very well wreck a BUNCH of other stuff. Let's not do that. 

> Now, here's the kicker. If the mod is looking for a DLC file and can't find it because it's in a packed DLC, it'll ignore it. There'll be a message in the DebugWindow, but otherwise, nada.

Back in 2013, the mod files I made for TMv1.0 replaced files inside the DLC SFAR. This, I am absolutely not wrong about, b/c it was a giant pain in the ass, and was why WV built a modmaking function into DLCEd2. Has the ability to replace inside the SFAR been removed from ModMaker? I think it's fine if it has, I just need to know for the wiki.

> There's no real way to display a message across all tools properly, but if the single DLC update is fixed, this won't be a concern.

I wasn't asking to display a message across all tools. I was saying that ModMaker should display the message. I was drawing a similarity between it and Texplorer to give you an example of what I meant.

---

Okay, all that said, let's talk about how to fix this.

**1. DLC Extraction.**
It sounds like the easiest thing is to make DLC extraction a requirement for ModMaker, just as it is for Texplorer. That way, everything is extracted and TOCing can stay working just as it is.

_My only point was that users need to be notified and given a choice prior to the extraction._ This is exactly how Texplorer handles things -- or at least used to. I haven't tested treebuilding since r727, so if you've since removed the notification and chance to abort, I'll end up posting another issue about that.

**2. TOCing.**
If the user is warned in advance and authorized DLC extraction, then TOCing all DLC -- even when not modified -- is no longer an issue. The user knows, it's not a surprise, and the TOCing can't hurt anything, as long as it's via autoto.

If we go this route, then it sounds like we should implement a "setup" procedure for ModMaker, similar to Texplorer. First time it opens, it prompts a message:

> FIRST TIME SETUP: ModMaker requires all DLC be unpacked to function. All DLC still packed in their original SFARs are about to be unpacked. The more DLC you have the longer it will take and the more hard drive space will be required. Click Okay to accept or Cancel to abort. If you abort, you will not be able to use ModMaker.

So... how about that?

**Comment by KFreon** [comment]
The problem with TOC'ing at the time was it wasn't working for whatever reason. Fob found out that rebuilding worked, and there's a comment there from him about it being a hack, but that it works.
I recall not having the time to figure out why the standard single DLC TOC update didn't work so I left it.

I am absolutely saying that this should be the functionality of all three of (my) tools. I don't want everything being extracted/rebuild in TPFTools if you change one DLC.
This is fortunate, because it can't be done any other way :)

**Replacing in DLC:** To my knowledge, there's no functions currently in Modmaker that allow this. Old mods MIGHT do this, but they might not. The ability remains in the code, but isn't used.

I really, reeeeally don't want to force Modmaker to require a tree/extracted DLC's. It's perfectly functional without the DLC's being extracted.

All in all, I plan on making the TOC change because that makes the most sense anyway. It's how it was before, it's how it should be. I just need to make it work :)
Should that come in the way of releasing the stable, I'll revert back to Fobs method (tiny change, takes about 5 seconds), and add a message to starting Modmaker, then go sob in the corner for a bit :P

**Comment by ghost** [comment]
Okay, at this point I'm totally confused. I'm not clear on how you want TPF Tools, ModMaker, and Texplorer to work in relation to DLC extraction and TOCing.

It would be good if you could post on the toolset forums in their designated threads and explain the changes. That way modders can jump into the conversation and point out if there are problems with whatever your strategy is. Doing too much development here on Git isn't a good idea; most folks are not reading this.

I'll  wait and see how ModMaker handles this in the next rev, but it's imperative that:
1. Users get notified and have a choice _if_ we're going to extract their DLC.
2. Changes to TOCing in ModMaker don't completely bork Texplorer when it comes to treebuilding, texture replacing, and using BW DLC + DLC mods.

**Re: Replacing in SFARs** -- I can verify that it broke somewhere along the way. 

If I load a TMv1.0 .mod file into r733 MM, autofix cannot fix it. For TMv1.0, I used .mod files to do full-PCC replacement inside the EC SFAR. Then the SFAR needed to be rebuilt and TOC'd with the DLC TOC.bin Updater to stay under the 2GB limit. This was back in 2013, before we were extracting SFARs to fix the 2GB limitation.

Now that SFARs can be extracted, the above method is deprecated, and it's far inferior anyway. Extracting the SFAR is a _much_ better way to go about it. It makes all DLC much easier to manage for the user. Files can easily be visualized and swapped in and out, just like the base game. Plus, of course, no 2GB limitation.

**Comment by KFreon** [comment]
This is what's confusing me though gift, there are no changes here.
All the tools have worked the way I mentioned above since they were made, with the extract and rebuild everything coming in recently to fix a weird bug I didn't have motivation to fix.

So I'll just outline how the three tools work now, and how I plan to make them work (like they used to)
**Texplorer**
The obvious one. It asks whether you want DLC extracted or not.
_Before and future:_  It used to (and should) only update the relevant DLC's that were affected by a texture change.

_Currently:_  It presently extracts and rebuild every DLC upon any kind of DLC change due to a bug I can't remember. Something to do with DLC TOC rebuild.

**TPFTools**
This has always used the Texplorer tree and done exactly what Texplorer does. It's literally a more user friendly interface. Underneath, it works the same.

**Modmaker**
The shunned outcast :P Seriously though, it does whatever the mod tells it to. It's why I don't like it.
If the mod says "edit this DLC", it'll try. If it can't, it bails like if it tried to edit "frankinstien.pcc" and obviously couldn't find it.
**Before and future:**  If it can, it does, then tells Texplorer to update TOC's (doesn't matter that it's the texture tool, it just calls the easy functions). Then since it's calling through Texplorer, it only updates the things it needs to.

**Currently:**  As with Texplorer, it currently extracts and updates all DLC's if anything touches a DLC.

TLDR: Texplorer is the only thing that will do ANYTHING about extracting DLC. The other tools do what it does, or ignore an unextracted DLC in Modmakers case.

Replacing in SFAR's is absolutely deprecated. I don't touch that anymore at all. I know I said mods can still do it, but Modmaker will tell users it's old, and if users ignore it, it's their own fault.
Seems rough, but there's a better way for a reason.

**Comment by ghost** [comment]
@KFreon -- Let's try to voice chat about this on Skype. 

It's a complex situation and we might be having certain communication issues on top of that. Guessing we can iron eveything out in 15-30 minutes, rather than days of back and forth, lol. I'll leave you an IM :)

**Comment by ghost** [comment]
@KFreon -- We should be talking about this stuff on Skype in a couple hours, but I had reached out to Fob previously, to make sure he was following things, since he/Heff wrote the current code for TOCing.

He's having PC issues atm, so he can't actually log into Git. He's asked if I can post some info on his behalf, so here it goes :)

---

**FOB SAYS:**

> "Any on screen messages by ModMaker about impending DLC extraction need to be less ambiguous, to avoid confusing players who already have DLC extracted." -- giftfish

This should be easy to do, changing messages is not a big deal. I guess we'd kjust have to work out what needs to be changed. Are you talking about the debug window or the message boxes or both? I haven't used the tools in a while so I am not sure what exactly the messages are like noe but I actually thought they were fairly good in the last version I used extensively (which something around 690).

> "ModMaker _should definitely ask_ permission before it extracts all DLC... especially if a DLC isn't even modded by a .mod file." -- giftfish

Hmmmm, not sure about this one. The problem is that I am not entirely sure how difficult it is to read out which DLCs require changes and which don't without reading out all the pcc data (and if you have to read it out, it might take just as long as extracting does in the first place). So I am not sure if K has a good reason to want to extract them all outright.

Also, I remember when I set up the new extraction process for textures, I did also set it up to extract every DLC automatically before the treescan. The thing is, there really is no harm in extracting them unless a user is really tight on HD space. It takes some time to be done but once done, it's just much more flexible and toolset friendly and we'll have much less issues in the code because we can always be sure that DLCs are extracted when we do modding operations. If we don't do the automatic extraction, we'd have to include tests in pretty much every other function that might need to change DLCs. This seems a lot of effort (with a huge potential to introduce bugs) just to prevent a procedure that is harmless to begin with and that only needs to be done once.

> "Akin to the above, ModMaker should only TOC DLC if it's been extracted. All DLC may be TOCed, but only if the user is okay with extracting them all." -- giftfish

I also wrote it like that on purpose. I'd leave it like it is. re-TOCing all DLCs after every modding operation (no matter if textures or .mods) is a good thing. In principle, you can't be TOCing too much, so the TOCing function should be run after everything that the toolset does really. Because again, it doesn't do any harm but it can prevent a whole slew of problems. It doesn't even take time, once DLCs are extracted, re-TOCing all of them takes less than 1 sec. Since I think it would be good that all DLCs are extracted in the very beginning of using the toolset in any way, the no-time scenario will be the case for the final TOCing always.

Basically the mantra is: There cannot be enough reTocing. ReTOC everything, all the time, every time. ReTOCing is good for the soul (well, our souls actually because if I never get a help message complaining about infinite loading screens again, it won't be too soon). :)

If anything (and to make the whole thing a little easier for users to comprehend), I would do the following when you first start ME3Explorer -- before even starting a tool -- there should be a message saying:

> Setting up game path ... auto-detecting ME3 game path...
> Your installation of ME3 is located in the following directory:
> <insert game dir, e.g. C:\Mass Effect 3\>
> Is this correct? <Input: Yes/No>"

If user chooses "No", he go to the "Set game path manually" dialogue.

After that, it should say:

> "ME3Explorer will now extract your game DLCs from their archives to make the game compatible with the toolset and prepare it for installation of mods. This may take a while, depending on your computer speed but it only needs to be done once.
> HD space needed to extract DLC: <Number of MB needed>
> Available space on game HD: <Number of MB of free space>
> Start DLC Extraction now?
> User input: <Yes (recommended)/No (advanced users only)>"

If the user chooses No, they'll get a warning:

> "Not extracting DLCs will make installation of many mods via this toolset impossible. If you want to extract your DLCs at a later time, use:
> Developer Tools -> DLC Tools -> DLC Editor 2 -> File -> Extract All DLCs"

That's it, if we do this at the very beginning, we are good for everything that comes later and don't have to worry about unextracted DLCs anymore. At the very least, this should happen when trying to use and texture/TPF tool or .mod tool.

Let me know what you/K think.

---

I really like this idea of triggering one instance of the path detection and DLC extraction, upon first opening the toolset. That way, it happens one time, outside of the context of any tool and it's done. Let's chat about this option :)

**Comment by MrFob** [comment]
Heyhey, I found my login details after all, so I thought I should post here.
Just a few additions I noticed when reading through this once more:

@K: In principle, there should be no problem just TOCing one DLC. If you look at my function for the "Extract all DLCs" class in DLC Editor 2, you'll see that it's just a simple for loop that does it for all DLCs. If you take that out, you can do it with single DLCs.

BUT: I strongly recommend we always TOC everything. I just wrote this to gift but I'll repeat it here:
IMO, the best way to make the toolset consistent is to have everyone who uses it extract all their DLC. This way, we can be sure that all DLCs are always extracted and we can always reTOC everything after any modding is done.
Advantage:
- No more TOC.bin issues
- No need to test for extracted/non-extracted DLCs in different functions (treescan, texture replacement, .mod installs, whatever)
- No need to distinguish between BioWare DLCs and mod DLCs because they will all be identical in structure

Disadvantage:
-The user has to go through the time consuming extract just once when using the tool for the first time after installing ME3

That's it, that's the only disadvantage I can see and IMO it's a minor thing given all the benefits.

That said, I am not current on TPF tool / modmaker code, so maybe I missed something.

As for the messages I suggested, that's just a very rough idea if what they could say, I am sure there is better wording for it.
By the way, it didn't copy properly but I suggested in the second message when the user can decide whether or not to extract DLCs right away, the buttons should say "yes (recommended)" and "No (advanced users only)" to make it really crystal clear that this is required.

Also, I made one mistake, these messages should not just show up when running the toolset for the first time. The problem here is that users may re-install the game but not the toolset. So on every startup, the toolset should check if there are any sfars that are larger than a few kb in the ME3 folder. If so it should say 
"It looks like your installation of ME3 has unextracted DLC. ME3Explorer will now extract your game DLCs from their archives to make the game compatible with yaddayaddayadda ..." see above.

**Comment by KFreon** [comment]
OK so as per skype chats (gift), here's what's going to happen.

There's a new shiny dialog that pops up at toolset start which asks if you want to extract DLC. Doesn't say how many, or tell you what they are. Just how much space. There's an abort button for us "advanced" users if we want to tempt fate.

This thing sits around extracting DLC in the background and you should be able to use most tools in the toolset while it's doing that. Modmaker, Texplorer, TPFTools, and some DLC stuff are disabled till it's done.

If all DLC is already extracted (checks by looking for XXX.bin in DLC folders), it doesn't prompt at all.

All tools don't make any mention of DLC/TOC etc. Texplorer still has it's FirstTimeSetup to allow DLC exclusion from the scan.

**Comment by ghost** [comment]
This sounds great. Can't wait to test.

**Comment by MrFob** [comment]
Perfect! I think that should take care of the entire messy DLC extraction thing once and for all.

Thanks KFreon!!!

**Comment by KFreon** [comment]
Thanks for the assist and info Fob and gift :D

**Comment by ghost** [comment]
Should be testing this today or tomorrow, btw :)

**Comment by ghost** [comment]
Issue resolved in rev746 (actually before).

ModMaker is still TOCing the base game when it doesn't need to, as reported in #258.

**Comment by ghost** [comment]
This seems the best issue to address this in. See #286 as to why I'm re-opening this temporarily.

@KFreon -- This is regarding the lock out you sneaked in for MM, TPF Tools, and TExplorer, when DLC extraction is turned off. We fought each other on this hard, but here's what I think we can do at this point:
1. **TPF Tools.** This doesn't need to be locked out (which I actually said originally). It's 100% dependent upon having trees in place, so Texplorer will need to be fiddled with a bit more.
2. **ModMaker.** We now have a dialog in place for when MM detects _any_ DLC that aren't extracted. Since extraction is all or nothing, this is fine. I think it's safe to keep this tool enabled.
3. **Texplorer.** This is the hardest one. I think it's safe to enable it, IF we implement a similar system to ModMaker. We need to add the ability for it to differentiate between packed and unpacked DLC (this uses TOC.bin presence/absence.). When a DLC is still packed, then it should by default be unchecked in the TexplorerFirstTimeSetup GUI. It should be grayed out and unclickable. You could even make a little mouseover that says "DLC not extracted" just to give folks a heads up :D

If we put this system in place for Texplorer, then it should be safe to leave it active, even when there's no extraction. It would confine modding textures to the vanilla game, but TOCing would still work, as it would just skip over TOCing any DLC still packed.

**Comment by ghost** [comment]
Verified in r748:
- K's tools are no longer locked upon extraction disable.
- Unextracted DLC are disabled in Texplorer (this is awesome!)

I think we are _finally_ done with this issue :dancers:
